columnflow/tasks/cms/external.py

```diff
@@ -0,0 +1,151 @@
+        mc_profile = self.read_mc_profile_from_cfg(externals.files.pu.mc_profile)
+
+        # loop over nominal and minbias_xs shifts
+        for shift in ["nominal", "minbias_xs_up", "minbias_xs_down"]:
+            # read or create the data profile
+            if self.data_mode == "hist":
+                pu_hist_target = externals.files.pu.data_profile[shift]
+                data_profile = self.read_data_profile_from_hist(pu_hist_target)
+            else:
+                pu_file_target = externals.files.pu.json
+                minbiasxs = self.config_inst.x.minbiasxs.get(shift)
+                data_profile = self.read_data_profile_from_pileupcalc(pu_file_target, minbiasxs)
+
+            # build the ratios and save them
+            if len(mc_profile) != len(data_profile):
+                raise Exception(
+                    f"the number of bins in the MC profile ({len(mc_profile)}) does not match that "
+                    f"of the data profile for the {shift} shift ({len(data_profile)})",
+                )
+
+            # store them
+            weights[shift] = [safe_div(d, m) for m, d in zip(mc_profile, data_profile)]
+            self.publish_message(f"computed pu weights for shift {shift}")
+
+        # save it
+        self.output().dump(weights, formatter="json")
+
+    @classmethod
+    def read_mc_profile_from_cfg(
+        cls,
+        pu_config_target: law.FileSystemTarget,
+    ) -> list[float]:
+        """
+        Takes a mc pileup configuration file stored in *pu_config_target*, parses its content and
+        return the pu profile as a list of float probabilities.
+        """
+        probs = []
+
+        # read non-empty lines
+        lines = map(lambda s: s.strip(), pu_config_target.load(formatter="text").split("\n"))
+
+        # loop through them and extract probabilities
+        found_prob_line = False
+        for line in lines:
+            # look for the start of the pu profile
+            if not found_prob_line:
+                if not line.startswith("mix.input.nbPileupEvents.probValue"):
+                    continue
+                found_prob_line = True
+                line = line.split("(", 1)[-1].strip()
+            # skip empty lines
+            if not line:
+                continue
+            # extract numbers
+            probs.extend(map(float, filter(bool, line.split(")", 1)[0].split(","))))
+            # look for the end of the pu profile
+            if ")" in line:
+                break
+
+        return cls.normalize_values(probs)
+
+    @classmethod
+    def read_data_profile_from_hist(
+        cls,
+        pu_hist_target: law.FileSystemTarget,
+    ) -> list[float]:
+        """
+        Takes the pileup profile in data preproducd by the lumi pog and stored in *pu_hist_target*,
+        builds the ratio to mc and returns the weights in a list.
+        """
+        hist_file = pu_hist_target.load(formatter="uproot")
+        probs = hist_file["pileup"].values().tolist()
```

Comment:
Why `to_list()`? The return type of `uproot.Histogram.values()` is a `numpy.ndarray`, right? Shouldn't this also suffice here (and also profit from numpy's power in general)?

Reply:
This is just to be in line with what `read_mc_profile_from_cfg` does, i.e., returning a list of floats that are passed to `normalize_values`. In total, there are 100 values, in the hists, so there shouldn't be too big of a performance loss :)